File: v0.2/moneytrace/core/engine.py
"""
Ledger engine — source of truth for all financial calculations.

RULES (from idea.md - NON-NEGOTIABLE):
- All amounts are integers in minor units (paise)
- Budget impact happens exactly once
- Engine is pure: no DB, no IO, no formatting
- Settlements never double-count budget
- Receivables excluded from budget by default
"""

from collections import defaultdict
from datetime import date
from typing import Iterable

from .events import EventType


# ---------------------------------------------------------------------------
# Core Budget Computation
# ---------------------------------------------------------------------------

def compute_available_budget(
    base_budget: int,
    events: Iterable[dict],
) -> int:
    """
    Compute remaining available budget.

    Formula (from idea.md):
        Available Budget = Base Monthly Budget
                         + Sum(BudgetAdjustments)
                         - Sum(Cash Expenses)
                         - Sum(Outstanding Liabilities)

    Args:
        base_budget: Monthly budget in minor units (paise)
        events: List of event dicts with 'type' and 'amount'

    Returns:
        Available budget in minor units
    """
    budget = base_budget

    for e in events:
        etype = e["type"]
        amount = e["amount"]

        if etype == EventType.EXPENSE:
            # Money spent -> budget down
            budget -= amount

        elif etype == EventType.LIABILITY:
            # Money reserved -> budget down
            budget -= amount

        elif etype == EventType.SETTLEMENT_RECEIVED:
            # Payment received -> budget relief
            budget += amount

        elif etype == EventType.BUDGET_ADJUSTMENT:
            # Extra money -> budget up
            budget += amount

        # RECEIVABLE -> no budget impact (informational)
        # SETTLEMENT_PAID -> no budget impact (already counted in LIABILITY)

    return budget


# ---------------------------------------------------------------------------
# Monthly Cash Spend
# ---------------------------------------------------------------------------

def compute_monthly_spend(
    events: Iterable[dict],
    month: int,
    year: int,
) -> int:
    """
    Cash that actually left your wallet in a month.

    Includes:
        - Expenses (direct spending)
        - Settlement paid (paying back liabilities)

    Excludes:
        - Liabilities (no cash out yet)
        - Receivables (no cash movement)
    """
    spend = 0

    for e in events:
        d: date = e["event_date"]
        if d.month != month or d.year != year:
            continue

        if e["type"] in (EventType.EXPENSE, EventType.SETTLEMENT_PAID):
            spend += e["amount"]

    return spend


# ---------------------------------------------------------------------------
# Outstanding Balances
# ---------------------------------------------------------------------------

def compute_outstanding_liabilities(events: Iterable[dict]) -> int:
    """
    Total amount you owe to others.

    LIABILITY creates debt, SETTLEMENT_PAID clears it.
    """
    total = 0

    for e in events:
        if e["type"] == EventType.LIABILITY:
            total += e["amount"]
        elif e["type"] == EventType.SETTLEMENT_PAID:
            total -= e["amount"]

    return max(total, 0)


def compute_outstanding_receivables(events: Iterable[dict]) -> int:
    """
    Total amount others owe you.

    RECEIVABLE creates credit, SETTLEMENT_RECEIVED clears it.
    """
    total = 0

    for e in events:
        if e["type"] == EventType.RECEIVABLE:
            total += e["amount"]
        elif e["type"] == EventType.SETTLEMENT_RECEIVED:
            total -= e["amount"]

    return max(total, 0)


# ---------------------------------------------------------------------------
# Friend Balances
# ---------------------------------------------------------------------------

def compute_friend_balances(events: Iterable[dict]) -> dict[str, int]:
    """
    Net balance per friend.

    Positive  -> friend owes you
    Negative  -> you owe friend
    """
    balances: dict[str, int] = defaultdict(int)

    for e in events:
        friend_id = e.get("friend_id")
        if not friend_id:
            continue

        amount = e["amount"]
        etype = e["type"]

        if etype == EventType.RECEIVABLE:
            # Friend owes you
            balances[friend_id] += amount

        elif etype == EventType.LIABILITY:
            # You owe friend
            balances[friend_id] -= amount

        elif etype == EventType.SETTLEMENT_RECEIVED:
            # Friend paid you back
            balances[friend_id] -= amount

        elif etype == EventType.SETTLEMENT_PAID:
            # You paid friend back
            balances[friend_id] += amount

    return dict(balances)


# ---------------------------------------------------------------------------
# Category Spending
# ---------------------------------------------------------------------------

def compute_category_spend(
    events: Iterable[dict],
    month: int | None = None,
    year: int | None = None,
) -> dict[str, int]:
    """
    Category-wise spending.

    Only EXPENSE events count (from idea.md: Settlement has NO category).
    """
    totals: dict[str, int] = defaultdict(int)

    for e in events:
        if e["type"] != EventType.EXPENSE:
            continue

        # Filter by month/year if specified
        if month and year:
            d: date = e["event_date"]
            if d.month != month or d.year != year:
                continue

        category = e.get("category") or "Uncategorized"
        totals[category] += e["amount"]

    return dict(totals)

File: mobile/lib/core/engine.dart
/// Ledger engine — source of truth for all financial calculations.
///
/// RULES:
/// - All amounts are integers in paise (minor units)
/// - Budget impact happens exactly once
/// - Engine is pure: no DB, no IO, no formatting
/// - Settlements never double-count budget
/// - Receivables excluded from budget by default

import 'events.dart';

/// Compute remaining available budget.
///
/// Formula:
///   Available = Base + Adjustments + Settlements Received
///             - Expenses - Liabilities - EMI Payments
int computeAvailableBudget(int baseBudget, Iterable<Map<String, dynamic>> events) {
  int budget = baseBudget;

  for (final e in events) {
    final etype = e['type'] as String;
    final amount = e['amount'] as int;

    if (etype == EventType.expense.value) {
      budget -= amount;
    } else if (etype == EventType.liability.value) {
      budget -= amount;
    } else if (etype == EventType.settlementReceived.value) {
      budget += amount;
    } else if (etype == EventType.budgetAdjustment.value) {
      budget += amount;
    } else if (etype == EventType.emiPayment.value) {
      budget -= amount;
    }
    // RECEIVABLE, SETTLEMENT_PAID, INCOME, TRANSFER -> no budget impact
  }

  return budget;
}

/// Compute total amount reserved for unpaid recurring transactions.
int computeUnpaidCommitments(Iterable<Map<String, dynamic>> unpaidRecurring) {
  int total = 0;

  for (final rec in unpaidRecurring) {
    final recType = rec['type'] as String? ?? '';
    final amount = rec['amount'] as int? ?? 0;

    if (recType == 'expense' || recType == 'emi_payment') {
      total += amount;
    }
  }

  return total;
}

/// Cash that actually left your wallet in a month.
int computeMonthlySpend(
  Iterable<Map<String, dynamic>> events,
  int month,
  int year,
) {
  int spend = 0;

  for (final e in events) {
    final d = DateTime.parse(e['event_date'] as String);
    if (d.month != month || d.year != year) continue;

    final etype = e['type'] as String;
    if (etype == EventType.expense.value ||
        etype == EventType.settlementPaid.value) {
      spend += e['amount'] as int;
    }
  }

  return spend;
}

/// Total amount you owe to others.
int computeOutstandingLiabilities(Iterable<Map<String, dynamic>> events) {
  int total = 0;

  for (final e in events) {
    final etype = e['type'] as String;
    if (etype == EventType.liability.value) {
      total += e['amount'] as int;
    } else if (etype == EventType.settlementPaid.value) {
      total -= e['amount'] as int;
    }
  }

  return total < 0 ? 0 : total;
}

/// Total amount others owe you.
int computeOutstandingReceivables(Iterable<Map<String, dynamic>> events) {
  int total = 0;

  for (final e in events) {
    final etype = e['type'] as String;
    if (etype == EventType.receivable.value) {
      total += e['amount'] as int;
    } else if (etype == EventType.settlementReceived.value) {
      total -= e['amount'] as int;
    }
  }

  return total < 0 ? 0 : total;
}

/// Days from `now` to the next occurrence of the EMI day-of-month.
/// Returns 0 if today *is* the EMI day. Clamps `emiDay` to a safe value
/// for short months (e.g. 31 in February becomes the last day of February).
int daysUntilNextEmi({required int emiDay, required DateTime now}) {
  final today = DateTime(now.year, now.month, now.day);
  final daysInThisMonth = DateTime(now.year, now.month + 1, 0).day;
  final clampedThisMonth = emiDay.clamp(1, daysInThisMonth);
  final dueThisMonth = DateTime(now.year, now.month, clampedThisMonth);

  if (!dueThisMonth.isBefore(today)) {
    return dueThisMonth.difference(today).inDays;
  }

  final nextMonth = DateTime(now.year, now.month + 1, 1);
  final daysInNextMonth = DateTime(nextMonth.year, nextMonth.month + 1, 0).day;
  final clampedNext = emiDay.clamp(1, daysInNextMonth);
  final dueNextMonth = DateTime(nextMonth.year, nextMonth.month, clampedNext);
  return dueNextMonth.difference(today).inDays;
}

/// Net balance per friend. Positive = friend owes you, Negative = you owe friend.
/// Reads only the legacy primary `friend_id` on events because that's the field
/// LIABILITY/RECEIVABLE/SETTLEMENT semantics depend on. Multi-friend tags from
/// the event_friends join table are *history* labels with no balance meaning.
Map<String, int> computeFriendBalances(Iterable<Map<String, dynamic>> events) {
  final balances = <String, int>{};

  for (final e in events) {
    final friendId = e['friend_id'] as String?;
    if (friendId == null) continue;

    final amount = e['amount'] as int;
    final etype = e['type'] as String;

    balances[friendId] = (balances[friendId] ?? 0);

    if (etype == EventType.receivable.value) {
      balances[friendId] = balances[friendId]! + amount;
    } else if (etype == EventType.liability.value) {
      balances[friendId] = balances[friendId]! - amount;
    } else if (etype == EventType.settlementReceived.value) {
      balances[friendId] = balances[friendId]! - amount;
    } else if (etype == EventType.settlementPaid.value) {
      balances[friendId] = balances[friendId]! + amount;
    }
  }

  return balances;
}

/// Compute the balance impact of an event on an account.
/// Returns negative for outflows, positive for inflows, 0 for no impact.
int balanceImpact(String type, int amount) {
  switch (type) {
    case 'expense':
    case 'settlement_paid':
    case 'credit_card_payment':
    case 'emi_payment':
      return -amount;
    case 'income':
    case 'settlement_received':
      return amount;
    default:
      return 0;
  }
}

/// Category-wise spending. Only EXPENSE events count.
Map<String, int> computeCategorySpend(
  Iterable<Map<String, dynamic>> events, {
  int? month,
  int? year,
}) {
  final totals = <String, int>{};

  for (final e in events) {
    if (e['type'] != EventType.expense.value) continue;

    if (month != null && year != null) {
      final d = DateTime.parse(e['event_date'] as String);
      if (d.month != month || d.year != year) continue;
    }

    final category = (e['category'] as String?) ?? 'Uncategorized';
    totals[category] = (totals[category] ?? 0) + (e['amount'] as int);
  }

  return totals;
}
